From: Olaf Dabrunz <od@suse.de>
Subject: [apm] default to "power_off" when SMP kernel is used on single processor machines
Reference: SUSE221667

This patch turns on support for the APM power_off function by default when the
SMP kernel is used on single processor machines.

Signed-off-by: Olaf Dabrunz <od@suse.de>


 arch/i386/kernel/apm.c |   14 +++++++++++++-
 1 files changed, 13 insertions(+), 1 deletion(-)

Index: arch/i386/kernel/apm.c
===================================================================
--- arch/i386/kernel/apm.c.orig	2007-01-15 14:41:28.000000000 +0100
+++ arch/i386/kernel/apm.c	2007-01-15 20:27:17.000000000 +0100
@@ -396,6 +396,7 @@ static int			smp __read_mostly;
 static int			apm_disabled = -1;
 #ifdef CONFIG_SMP
 static int			power_off;
+static int __initdata		power_off_set;
 #else
 static int			power_off = 1;
 #endif
@@ -1868,8 +1869,10 @@ static int __init apm_setup(char *str)
 		if (strncmp(str, "debug", 5) == 0)
 			debug = !invert;
 		if ((strncmp(str, "power-off", 9) == 0) ||
-		    (strncmp(str, "power_off", 9) == 0))
+		    (strncmp(str, "power_off", 9) == 0)) {
 			power_off = !invert;
+			power_off_set = 1;
+		}
 		if (strncmp(str, "smp", 3) == 0)
 		{
 			smp = !invert;
@@ -1892,6 +1895,15 @@ static int __init apm_setup(char *str)
 }
 
 __setup("apm=", apm_setup);
+
+#ifdef CONFIG_SMP
+static void __init apm_poweroff_init(void)
+{
+	if (!power_off_set)
+		power_off = (num_online_cpus() == 1);
+}
+late_initcall(apm_poweroff_init);
+#endif
 #endif
 
 static struct file_operations apm_bios_fops = {
